File: server/models/Notification.ts
import mongoose from 'mongoose';

enum NotificationTypeEnum {
  info = 'info',
  followRequest = 'follow-request'
}

interface NotificationData {
  user?: mongoose.Schema.Types.ObjectId;
  post?: mongoose.Schema.Types.ObjectId;
}

interface Notification {
  content: string;
  user: mongoose.ObjectId;
  data: NotificationData;
  isRead: boolean;
  type: NotificationTypeEnum;
  createdAt: Date;
  updatedAt: Date;
}

const dataSchema = new mongoose.Schema<NotificationData>({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: false },
  post: { type: mongoose.Schema.Types.ObjectId, ref: 'Post', required: false }
});

const schema = new mongoose.Schema<Notification>(
  {
    content: { type: String, required: true, maxlength: 100 },
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    data: { type: dataSchema, required: false },
    isRead: { type: Boolean, default: false },
    type: { type: String, enum: Object.values(NotificationTypeEnum) }
  },
  { toJSON: { virtuals: true }, timestamps: true }
);

const Notification = mongoose.model<Notification>('Notification', schema);
type NotificationType = mongoose.HydratedDocument<Notification>;

export { NotificationType, Notification, NotificationTypeEnum };

File: mobile/lib/models/notification.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:instagram_clone/models/converters.dart';
part 'notification.freezed.dart';
part 'notification.g.dart';

enum NotificationType {
  @JsonValue('follow-request')
  followRequest,
  @JsonValue('info')
  info
}

@freezed
class GetNotificationsResponse with _$GetNotificationsResponse {
  const factory GetNotificationsResponse({
    required bool success,
    required List<GetNotificationsResponseData> data,
  }) = _GetNotificationsResponse;

  factory GetNotificationsResponse.fromJson(Map<String, Object?> json) =>
      _$GetNotificationsResponseFromJson(json);
}

@freezed
class NotificationDataUser with _$NotificationDataUser {
  const factory NotificationDataUser({
    required String id,
    required String avatar,
    required String username,
    required String name,
  }) = _NotificationDataUser;

  factory NotificationDataUser.fromJson(Map<String, Object?> json) =>
      _$NotificationDataUserFromJson(json);
}

@freezed
class NotificationData with _$NotificationData {
  const factory NotificationData({
    required NotificationDataUser? user,
    required String? post,
  }) = _NotificationData;

  factory NotificationData.fromJson(Map<String, Object?> json) =>
      _$NotificationDataFromJson(json);
}

@freezed
class GetNotificationsResponseData with _$GetNotificationsResponseData {
  const factory GetNotificationsResponseData({
    required String id,
    required String content,
    required NotificationData? data,
    required NotificationType type,
    @DateTimeConvertor() required DateTime createdAt,
    @DateTimeConvertor() required DateTime updatedAt,
  }) = _GetNotificationsResponseData;

  factory GetNotificationsResponseData.fromJson(Map<String, Object?> json) =>
      _$GetNotificationsResponseDataFromJson(json);
}
